spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/msi/ConfigMSICredentials.java

```diff
@@ -0,0 +1,196 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.StringUtils;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Managed Service Identity token based credentials for use with a REST Service Client.
+ *
+ * Modified from https://azure.github.io/azure-sdk-for-java/com/microsoft/azure/credentials/MSICredentials.html,
+ * to support api version configuration, and removed VM extension support which will be deprecated.
+ */
+public class ConfigMSICredentials extends AzureTokenCredentials {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigMSICredentials.class);
+
+    private static final String RETRIEVE_IMDS_FAILURE_MSG =
+            "Failed to retrieve token from Azure Instance Metadata Service.";
+    private static final int[] RETRY_SLOTS = new int[]{0, 2, 6, 14, 30}; /* Seconds */
+    private static final int MAX_RETRY = RETRY_SLOTS.length;
+    private static final CloseableHttpClient HTTP_CLIENT = HttpClients.createDefault();
+    private static final AzureJacksonAdapter ADAPTER = new AzureJacksonAdapter();
+
+    private final Lock lock = new ReentrantLock();
+    private final ConcurrentHashMap<String, MSIToken> cache = new ConcurrentHashMap<>();
+    private final String resource;
+    private AzureInstanceMetadataService metadataService;
+    private ImmutableMap<String, String> tokenReqHeaders = ImmutableMap.of();
+    private MSIType msiType;
+
+    public ConfigMSICredentials() {
+        this(AzureEnvironment.AZURE);
+    }
+
+    /**
+     * Initializes a new instance of the ConfigMSICredentials.
+     *
+     * @param environment the Azure environment to use
+     */
+    public ConfigMSICredentials(AzureEnvironment environment) {
+        super(environment, null); /* retrieving MSI token does not require tenant */
+        this.resource = environment.managementEndpoint();
+        this.msiType = checkMSIType();
+        this.metadataService = getMetadataService(msiType);
+        setHeader(msiType);
+    }
+
+    @Override
+    public String getToken(String resource) {
+        return this.getTokenFromIMDSEndpoint(resource == null ? this.resource : resource);
+    }
+
+    private String getTokenFromIMDSEndpoint(String resource) {
+        MSIToken token = cache.get(resource);
+        if (token != null && !token.isExpired()) {
+            return token.accessToken();
+        }
+        lock.lock();
+        try {
+            token = cache.get(resource);
+            if (token != null && !token.isExpired()) {
+                return token.accessToken();
+            }
+
+            token = retrieveTokenFromIDMSWithRetry(resource);
+            if (token != null) {
+                cache.put(resource, token);
+            }
+            return token.accessToken();
```

Comment:
Kind of confused. So is `token` will be `null`? If yes, this line will probably throw NPE?

Reply:
Good catch, reused some code and now the token should not be null if no exception was thrown.